py/usageaccountant/datadog_fetcher.py

```diff
@@ -0,0 +1,292 @@
+import argparse
+import json
+import logging
+import urllib.parse
+from typing import Any, Dict, List, Optional, Tuple, TypedDict
+from urllib.request import Request, urlopen
+
+from usageaccountant.accumulator import (
+    KafkaConfig,
+    UsageAccumulator,
+    UsageUnit,
+)
+
+# TODO fetch these from env
+headers = {"DD-APPLICATION-KEY": "", "DD-API-KEY": ""}
+
+logger = logging.getLogger("fetcher")
+
+# TODO remove this before prod
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(asctime)s [%(levelname)s] %(message)s",
+    handlers=[logging.FileHandler("debug.log"), logging.StreamHandler()],
+)
+
+
+class DatadogResponseUnit(TypedDict):
+    plural: str
+    name: str
+
+
+class DatadogResponseSeries(DatadogResponseUnit):
+    unit: List[DatadogResponseUnit]
+    pointlist: List[List[float]]
+    scope: str
+
+
+class DatadogResponse(DatadogResponseSeries):
+    error: Optional[str]
+    query: str
+    from_date: int
+    to_date: int
+    status: str
+    series: List[DatadogResponseSeries]
+
+
+def is_valid_query(query: str) -> bool:
+    """
+    Validates if query string contains shared_resource_id
+    and app_feature–parameters used in UsageAccountant.record().
+
+    query: Datadog query
+    """
+    return "shared_resource_id" in query and "app_feature" in query
+
+
+def get(query: str, start_time: int, end_time: int) -> Any:
+    """
+    Fetches timeseries data from Datadog API, returns response dict.
+
+    query: Datadog query
+    start_time: Start of the time window in Unix epoch format
+                with second-level precision
+    end_time: End of the time window in Unix epoch format
+              with second-level precision
+    """
+    data = {"query": query, "from": start_time, "to": end_time}
+
+    base_url = "https://api.datadoghq.com/api/v1/query"
+    query_string = urllib.parse.urlencode(data)
+    full_url = f"{base_url}?{query_string}"
+
+    try:
+        request = Request(url=full_url, headers=headers)
+        # context is needed to handle
+        # ssl.SSLCertVerificationError:
+        # [SSL: CERTIFICATE_VERIFY_FAILED] on macOS
+        # TODO remove next three lines and
+        # context parameter from the call to urlopen()
+        import ssl
+
+        context = ssl.SSLContext()
+        response = urlopen(request, context=context).read()
+        return json.loads(response)
+
+    except Exception as e:
+        raise e
+
+
+def parse_response_series_and_unit(
+    response: DatadogResponse,
+) -> Tuple[List[DatadogResponseSeries], UsageUnit]:
+    """
+    Extracts list of series objects and unit from the API response.
+
+    response: response received from API call
+    """
+
+    error_msg = response.get("error")
+    if error_msg:
+        raise Exception(format_exception_message(response, error_msg))
+
+    if response.get("series") is not None:
```

Comment:
Can this be none in a successful response? 

If there is something wrong in your answer but you get a valid json and no error message, Each of the condition above would make the script silently fail at recording data.

You want to make it fail if you do not get a valid response from datadog. 
Or at least log errors if you get empty response.

Reply:
I think the last three [checks](https://github.com/getsentry/usage-accountant/blob/e8c3bda124dc39e9cc7c57e8636093402675ba0b/py/usageaccountant/datadog_fetcher.py#L134) should handle edge cases like the one you mentioned.

If not, can you please share a sample response that would still fail silently through those checks?

Reply:
So the best option is to first check for the invalid condition and raise if you encounter it rather than wait till the end, it is like inverting the condition, the code becomes less indented and the reader knows immediately what the invariant is that you look for.